src/python/WMComponent/WorkflowUpdater/SiteListPoller.py

```diff
@@ -0,0 +1,156 @@
+        self.localWQ = WorkQueue(self.localCouchUrl,
+                                 self.config.WorkQueueManager.dbname)
+
+    def getActiveWorkflows(self):
+        """
+        Provide list of active requests within WMAgent
+        :return: dict of workflows names vs pickle files
+        """
+        # get list of active workflows in WMAgent
+        wflowSpecs = self.listActiveWflows.execute()
+
+        # construct dictionary of workflow names and their pickle files
+        wmaSpecs = {}
+        for wflowSpec in wflowSpecs:
+            name = wflowSpec['name']  # this is the name of workflow
+            pklFileName = wflowSpec['spec']  # the "spec" in WMBS table (wmbs_workflow.spec) is pkl file name
+            wmaSpecs[name] = pklFileName
+        return wmaSpecs
+
+    def getRequestSpecs(self, requests):
+        """
+        Return list of requests specs for provided list of request names
+        :param requests: list of workflow requests names
+        :return wsList: list of workflow records obtained from wmstats server, each record has the following structure
+        {"RequestName": "bla", "SiteWhitelist":[], "SiteBlacklist": []}
+        """
+        # get list of workflows from wmstats
+        states = ['running-open', 'acquired']
+        outputMask = ['SiteWhiteList', 'SiteBlackList']
+        wsList = []
+        for state in states:
+            inputConditions = {"RequestStatus": state}
+            resp = self.wmstatsSrv.getFilteredActiveData(inputConditions, outputMask)
+            data = json.loads(resp['data'])
+            for rdict in data['result']:
+                # rdict here has the following structure: list of records where each record is
+                # {"RequestName": "bla", "SiteWhitelist":[], "SiteBlacklist": []}
+                wflow = rdict['RequestName']
+                # check that our workflow is in our requests list
+                if wflow in requests:
+                    wsList.append(rdict)
+        return wsList
+
+    @timeFunction
+    def algorithm(self, parameters=None):
+        """
+        Perform the following logic:
+        - obtain list of current active workflows from the agent
+        - requests their specs from upstream ReqMgr2 server
+        - update site lists of all workflows
+        - push new specs to the agent local WorkQueue and update pickle spec file
+
+        :return: none
+        """
+        # get list of active workflows from the agent, the returned dict
+        # is composed by workflow names and associated pickle file (data comes from WMBS)
+        wmaSpecs = self.getActiveWorkflows()
+        wflows = wmaSpecs.keys()
+
+        # obtain workflow records from wmstats server
+        wsList = self.getRequestData(wflows)
+
+        # iterate over workflow items and update local WorkQueue and pickle files if
+        # either site white or black lists are different
+        for rdict in wsList:
```

Comment:
We should make sure that we iterate over the list of active workflows, which is supposed to be much smaller than the list of active workflows in WMStats. Otherwise the cost of loading each spec file will be unnecessarily high.

Reply:
ok, changes are made.